app/javascript/components/patient/vaccines/VaccineModal.js

```diff
@@ -0,0 +1,244 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button, Modal, Row, Col, Form } from 'react-bootstrap';
+import Select from 'react-select';
+import moment from 'moment';
+import ReactTooltip from 'react-tooltip';
+
+import DateInput from '../../util/DateInput';
+
+class VaccineModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      product_name: this.props.currentVaccineData?.product_name,
+      // If there is not already a group name, default to the first option
+      group_name: this.props.currentVaccineData?.group_name
+        ? this.props.currentVaccineData.group_name
+        : this.props.group_name_options
+        ? this.props.group_name_options[0]
+        : null,
+      administration_date: this.props.currentVaccineData?.administration_date,
+      dose_number: this.props.currentVaccineData?.dose_number,
+      notes: this.props.currentVaccineData?.notes,
+      // Sorting so that the blank option shows up at the top
+      sorted_dose_number_options: this.props.dose_number_options ? this.props.dose_number_options.sort() : [],
+      isValid: false,
+    };
+  }
+
+  componentDidMount() {
+    this.checkIfValid();
+  }
+
+  /**
+   * Data is valid and can be saved if there is both a group name and product name selected.
+   */
+  checkIfValid = () => {
+    const isValid = this.state.group_name && this.state.product_name;
+    this.setState({ isValid });
+  };
+
+  handleProductNameChange = data => {
+    this.setState({ product_name: data.value }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleGroupNameChange = data => {
+    // Resets product name if selected because product name options are dependent on group name
+    this.setState({ group_name: data.value, product_name: null }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleAdministrationDateChange = newDate => {
+    this.setState({ administration_date: newDate }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleDoseNumberChange = data => {
+    this.setState({ dose_number: data.value }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleNotesChange = event => {
+    const val = event.target.value;
+    this.setState({ notes: val }, () => {
+      this.checkIfValid();
```

Comment:
same comment as above

Reply:
This code has been removed (updated this logic to be more concise).